llvm/lib/Object/SYCLBIN.cpp

```diff
@@ -0,0 +1,447 @@
+//===- SYCLBIN.cpp - SYCLBIN binary format support --------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Object/SYCLBIN.h"
+
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Object/OffloadBinary.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace llvm;
+using namespace llvm::object;
+
+namespace {
+
+class SYCLBINBlockReader {
+protected:
+  SYCLBINBlockReader(const char *Data, size_t Size) : Data{Data}, Size{Size} {}
+
+  Error ReadSizeCheck(size_t ByteOffset, size_t ReadSize) {
+    if (ByteOffset + ReadSize > Size)
+      return createStringError(inconvertibleErrorCode(),
+                               "Unexpected file contents size.");
+    return Error::success();
+  }
+
+  const char *Data = nullptr;
+  size_t Size = 0;
+};
+
+class SYCLBINHeaderBlockReader : public SYCLBINBlockReader {
+public:
+  SYCLBINHeaderBlockReader(const char *Data, size_t Size)
+      : SYCLBINBlockReader(Data, Size) {}
+
+  template <typename HeaderT>
+  Expected<const HeaderT *> GetHeaderPtr(size_t ByteOffset) {
+    if (Error ReadSizeError = ReadSizeCheck(ByteOffset, sizeof(HeaderT)))
+      return ReadSizeError;
+    return reinterpret_cast<const HeaderT *>(Data + ByteOffset);
+  }
+};
+
+class SYCLBINByteTableBlockReader : public SYCLBINBlockReader {
+public:
+  SYCLBINByteTableBlockReader(const char *Data, size_t Size)
+      : SYCLBINBlockReader(Data, Size) {}
+
+  Expected<StringRef> GetBinaryBlob(size_t ByteOffset, uint64_t BlobSize) {
+    if (Error ReadSizeError = ReadSizeCheck(ByteOffset, BlobSize))
+      return ReadSizeError;
+    return llvm::StringRef{Data + ByteOffset, BlobSize};
+  }
+
+  Expected<std::unique_ptr<llvm::util::PropertySetRegistry>>
+  GetMetadata(size_t ByteOffset, uint64_t MetadataSize) {
+    Expected<StringRef> BlobOrError = GetBinaryBlob(ByteOffset, MetadataSize);
+    if (!BlobOrError)
+      return BlobOrError.takeError();
+
+    std::string PropStr{BlobOrError->str()};
```

Comment:
nit.
We could create here a `MemoryBuffer` right from the `StringRef`.

Reply:
Sadly it won't work. `PropertySetRegistry::read` expects null-termination and the binary blob doesn't have that. It gets it when we copy it to a string though.